swap/src/bin/asb.rs

```diff
@@ -212,3 +232,46 @@ async fn init_monero_wallet(
 
     Ok(wallet)
 }
+
+/// Registers a hidden service for each network.
+/// Note: Once ac goes out of scope, the services will be de-registered.
+async fn register_tor_services(
+    networks: Vec<Multiaddr>,
+    uac: UnauthenticatedConnection,
+) -> Result<AuthenticatedConnection> {
+    let mut ac = uac.into_authenticated_connection().await?;
+
+    let hidden_services_details = networks
+        .iter()
+        .flat_map(|network| {
+            network.iter().map(|protocol| match protocol {
+                Protocol::Tcp(port) => Some((
+                    port,
+                    SocketAddr::new(IpAddr::from(Ipv4Addr::new(127, 0, 0, 1)), port),
+                )),
+                _ => {
+                    // We only care for Tcp for now.
+                    None
+                }
+            })
+        })
+        .filter_map(|details| details)
+        .collect::<Vec<_>>();
+
+    // Note: ideally we have deterministic keys but unfortunately there is no Api
+    // for that yet.
+    let key = TorSecretKeyV3::generate();
+
+    ac.add_services(&hidden_services_details, &key).await?;
+
+    let onion_address = key
+        .public()
+        .get_onion_address()
+        .get_address_without_dot_onion();
+
+    hidden_services_details.iter().for_each(|(port, _)| {
+        tracing::info!("/onion3/{}:{}", onion_address, port);
```

Comment:
Another benefit of moving the creation of ephemeral services into the Transport will be that we can "reuse" the log in the `EventLoop` about `NewListenAddr` as that will be automatically triggered if we report a [`ListenerEvent::NewAddress`](https://docs.rs/libp2p/0.37.1/libp2p/core/transport/enum.ListenerEvent.html#variant.NewAddress).